Min/max errors with visual

There is a header on windows, Windows.h, that defines maccros called min() and max(), causing warnings and errors in code when calling min or max. 
I get those with Visual 2017 :
```
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(130,41): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(158,49): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(172,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(175,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(178,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(182,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(185,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\matchBase.hpp(188,75): war
ning C4003: not enough arguments for function-like macro invocation 'max' [C:\C
ygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_with_
OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr/utils/registrationMetrics.h(35,52): w
arning C4003: not enough arguments for function-like macro invocation 'max' [C:
\Cygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_wit
h_OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr/utils/registrationMetrics.h(74,52): w
arning C4003: not enough arguments for function-like macro invocation 'max' [C:
\Cygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_wit
h_OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\match4pcsBase.hpp(58,60):
warning C4003: not enough arguments for function-like macro invocation 'max' [C
:\Cygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_wi
th_OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\match4pcsBase.hpp(154,69):
 warning C4003: not enough arguments for function-like macro invocation 'max' [
C:\Cygwin\home\afabri\cgal\Point_set_processing_3\examples\build\registration_w
ith_OpenGR.vcxproj]
C:\3rdPartyLibs\OpenGR\install\include\gr\algorithms\match4pcsBase.hpp(154,69):
 error C2589: '(': illegal token on right side of '::' [C:\Cygwin\home\afabri\c
gal\Point_set_processing_3\examples\build\registration_with_OpenGR.vcxproj]

```

To fix those, it is necessary to replace every occurence of `std::max()` by `(std::max)()`, 
`std::numeric_limits<Scalar>::max()` by `(std::numeric_limits<Scalar>::max)()`, etc...

--- a/src/gr/algorithms/PointPairFilter.h
+++ b/src/gr/algorithms/PointPairFilter.h
@@ -99,7 +99,7 @@ struct PairFilterConcept {
                 const double second_normal_angle = (q.normal() + p.normal()).norm();
                 // Take the smaller normal distance.
                 const Scalar first_norm_distance =
-                        std::min(std::abs(first_normal_angle - pair_normals_angle),
+                        (std::min)(std::abs(first_normal_angle - pair_normals_angle),
                                  std::abs(second_normal_angle - pair_normals_angle));
                 // Verify appropriate angle between normals and distance.
 

--- a/src/gr/algorithms/congruentSetExplorationBase.hpp
+++ b/src/gr/algorithms/congruentSetExplorationBase.hpp
@@ -160,7 +160,7 @@ CongruentSetExplorationBase<Traits, PointType, TransformVisitor, PairFilteringFu
         std::chrono::duration_cast<std::chrono::seconds>
         (system_clock::now() - t0).count() /
                           MatchBaseType::options_.max_time_seconds;
-    Scalar fraction = std::max(fraction_time, fraction_try);
+    Scalar fraction = (std::max)(fraction_time, fraction_try);
 
     if (v.needsGlobalTransformation()) {
       getGlobalTransform(transformation);

--- a/src/gr/algorithms/match4pcsBase.hpp
+++ b/src/gr/algorithms/match4pcsBase.hpp
@@ -55,7 +55,7 @@ namespace gr {
         typename PointType::Scalar &invariant2,
         int &id1, int &id2, int &id3, int &id4) {
 
-        Scalar min_distance = std::numeric_limits<Scalar>::max();
+        Scalar min_distance = (std::numeric_limits<Scalar>::max)();
         int best1, best2, best3, best4;
         best1 = best2 = best3 = best4 = -1;
         for (int i = 0; i < 4; ++i) {
@@ -151,7 +151,7 @@ namespace gr {
                 Scalar C =
                         (-x2 * y1 + x3 * y1 + x1 * y2 - x3 * y2 - x1 * y3 + x2 * y3) / denom;
                 base4 = -1;
-                Scalar best_distance = std::numeric_limits<Scalar>::max();
+                Scalar best_distance = (std::numeric_limits<Scalar>::max)();
                 // Go over all points in P.
                 const Scalar too_small = std::pow(MatchBaseType::max_base_diameter_ * kBaseTooSmall, 2);
                 for (unsigned int i = 0; i < MatchBaseType::sampled_P_3D_.size(); ++i) {

--- a/src/gr/algorithms/matchBase.hpp
+++ b/src/gr/algorithms/matchBase.hpp
@@ -127,7 +127,7 @@ MATCH_BASE_TYPE::ComputeRigidTransformation(const Coordinates& ref,
         bool computeScale ) const {
     static const Scalar pi = std::acos(-1);
 
-    rms_ = std::numeric_limits<Scalar>::max();
+    rms_ = (std::numeric_limits<Scalar>::max)();
 
     Scalar kSmallNumber = 1e-6;
 
@@ -155,7 +155,7 @@ MATCH_BASE_TYPE::ComputeRigidTransformation(const Coordinates& ref,
         const Scalar ratioMean = (ratio1+ratio2)/Scalar(2.);            // mean of the two
 
         if ( ratioDev > Scalar(0.1) )
-            return std::numeric_limits<Scalar>::max();
+            return (std::numeric_limits<Scalar>::max)();
 
 
         //Log<LogLevel::Verbose>( ratio1, " ", ratio2, " ", ratioDev, " ", ratioMean);
@@ -169,23 +169,23 @@ MATCH_BASE_TYPE::ComputeRigidTransformation(const Coordinates& ref,
     }
 
     VectorType vector_p1 = p1 - p0;
-    if (vector_p1.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_p1.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_p1.normalize();
     VectorType vector_p2 = (p2 - p0) - ((p2 - p0).dot(vector_p1)) * vector_p1;
-    if (vector_p2.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_p2.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_p2.normalize();
     VectorType vector_p3 = vector_p1.cross(vector_p2);
-    if (vector_p3.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_p3.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_p3.normalize();
 
     VectorType vector_q1 = q1 - q0;
-    if (vector_q1.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_q1.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_q1.normalize();
     VectorType vector_q2 = (q2 - q0) - ((q2 - q0).dot(vector_q1)) * vector_q1;
-    if (vector_q2.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_q2.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_q2.normalize();
     VectorType vector_q3 = vector_q1.cross(vector_q2);
-    if (vector_q3.squaredNorm() == 0) return std::numeric_limits<Scalar>::max();
+    if (vector_q3.squaredNorm() == 0) return (std::numeric_limits<Scalar>::max)();
     vector_q3.normalize();
 
     //cv::Mat rotation = cv::Mat::eye(3, 3, CV_64F);
@@ -291,7 +291,7 @@ void MATCH_BASE_TYPE::init(const InputRange1& P,
         std::vector<int> indices(uniform_Q.size());
         std::iota( std::begin(indices), std::end(indices), 0 );
         std::shuffle(indices.begin(), indices.end(), randomGenerator_);
-        size_t nbSamples = std::min(uniform_Q.size(), options_.sample_size);
+        size_t nbSamples = (std::min)(uniform_Q.size(), options_.sample_size);
         indices.resize(nbSamples);
 
         // using the indices, copy elements from uniform_Q to sampled_P_3D_

--- a/src/gr/utils/registrationMetrics.h
+++ b/src/gr/utils/registrationMetrics.h
@@ -32,7 +32,7 @@ namespace Utils{
 template <typename Scalar>
 struct LCPMetric {
     /// Support size of the LCP
-    Scalar epsilon_ = std::numeric_limits<Scalar>::max();
+    Scalar epsilon_ = (std::numeric_limits<Scalar>::max)();
 
     template <typename Range>
     inline Scalar operator()( const gr::KdTree<Scalar> & ref,
@@ -71,7 +71,7 @@ struct LCPMetric {
 template <typename Scalar>
 struct WeightedLCPMetric {
     /// Support size of the LCP
-    Scalar epsilon_ = std::numeric_limits<Scalar>::max();
+    Scalar epsilon_ = (std::numeric_limits<Scalar>::max)();
 
     template <typename Range>
     inline Scalar operator()( const gr::KdTree<Scalar> & ref,

